plugin/trino-hive/src/main/java/io/trino/plugin/hive/procedure/SyncPartitionMetadataProcedure.java

```diff
@@ -137,23 +138,48 @@ private void doSyncPartitionMetadata(ConnectorSession session, ConnectorAccessCo
                 accessControl.checkCanDeleteFromTable(null, new SchemaTableName(schemaName, tableName));
             }
 
-            Location tableLocation = Location.of(table.getStorage().getLocation());
-
-            Set<String> partitionsInMetastore = metastore.getPartitionNames(schemaName, tableName)
+            Set<String> canonicalPartitionNamesInMetastore = metastore.getPartitionNames(schemaName, tableName)
                     .map(ImmutableSet::copyOf)
                     .orElseThrow(() -> new TableNotFoundException(schemaTableName));
-            Set<String> partitionsInFileSystem = listPartitions(fileSystemFactory.create(session), tableLocation, table.getPartitionColumns(), caseSensitive);
+            String tableStorageLocation = table.getStorage().getLocation();
+            Set<String> actualPartitionNamesInMetastore = metastore.getPartitionsByNames(schemaName, tableName, ImmutableList.copyOf(canonicalPartitionNamesInMetastore))
+                    .values().stream()
+                    .flatMap(Optional::stream) // disregard partitions which disappeared in the meantime since listing the partition names
+                    .map(partition -> getPartitionNameFromPartitionLocation(partition.getStorage().getLocation(), tableStorageLocation))
+                    .collect(toImmutableSet());
+
+            Set<String> partitionsInFileSystem = listPartitions(fileSystemFactory.create(session), Location.of(tableStorageLocation), table.getPartitionColumns(), caseSensitive);
 
             // partitions in file system but not in metastore
-            Set<String> partitionsToAdd = difference(partitionsInFileSystem, partitionsInMetastore);
+            Set<String> partitionsToAdd = difference(partitionsInFileSystem, actualPartitionNamesInMetastore);
 
             // partitions in metastore but not in file system
-            Set<String> partitionsToDrop = difference(partitionsInMetastore, partitionsInFileSystem);
+            Set<String> partitionsToDrop = difference(actualPartitionNamesInMetastore, partitionsInFileSystem);
 
             syncPartitions(partitionsToAdd, partitionsToDrop, syncMode, metastore, session, table);
         }
     }
 
+    private static String getPartitionNameFromPartitionLocation(String partitionLocation, String tableLocation)
+    {
+        if (!partitionLocation.startsWith(tableLocation)) {
+            throw new IllegalArgumentException("Partition location '%s' does not belong to base table location '%s'".formatted(partitionLocation, tableLocation));
```

Comment:
Hive allows having partitions outside of the table location with `ALTER TABLE xxx ADD PARTITION (x=x) LOCATION '...'` statement. This change may cause a regression in such a case, right?

Reply:
Thank you for the heads'up.
I missed this detail :(

Reply:
I added logic to disregard nonconventional Hive partitions in the syncing logic.